File: go/pkg/vax/sdto/SchemaBuilder.go
package sdto

type SchemaBuilder struct {
	Actions map[string]FieldSpec
}

// 啟動點
func NewSchemaBuilder() *SchemaBuilder {
	return &SchemaBuilder{
		Actions: make(map[string]FieldSpec),
	}
}

// 設定行動字串長度限制
func (b *SchemaBuilder) SetActionStringLength(action string, min string, max string) *SchemaBuilder {
	b.Actions[action] = FieldSpec{
		Type: "string",
		Min:  &min,
		Max:  &max,
	}
	return b
}

// 設定行動數字範圍限制
func (b *SchemaBuilder) SetActionNumberRange(action string, min string, max string) *SchemaBuilder {
	b.Actions[action] = FieldSpec{
		Type: "number",
		Min:  &min,
		Max:  &max,
	}
	return b
}

// 設定行動列舉限制
func (b *SchemaBuilder) SetActionEnum(action string, values []string) *SchemaBuilder {
	b.Actions[action] = FieldSpec{
		Type: "string",
		Enum: values,
	}
	return b
}

// 支援的簽名類型
var SupportedSignTypes = []string{"ed25519", "rsa", "ecdsa"}

// 設定簽名欄位，指定簽名演算法類型
func (b *SchemaBuilder) SetActionSign(action string, signType string) *SchemaBuilder {
	b.Actions[action] = FieldSpec{
		Type: "sign",
		Enum: []string{signType},
	}
	return b
}

// 設定簽名欄位，允許多種簽名類型
func (b *SchemaBuilder) SetActionSignMulti(action string, signTypes []string) *SchemaBuilder {
	b.Actions[action] = FieldSpec{
		Type: "sign",
		Enum: signTypes,
	}
	return b
}

// BuildSchema 回傳給 constructor 用的 FieldSpec map
func (b *SchemaBuilder) BuildSchema() map[string]FieldSpec {
	return b.Actions
}

// Build 回傳 JSON 友善格式（跨語言傳輸用）
func (b *SchemaBuilder) Build() map[string]any {
	props := map[string]any{}

	for name, c := range b.Actions {
		m := map[string]any{
			"type": c.Type,
		}
		if c.Min != nil {
			m["min"] = *c.Min
		}
		if c.Max != nil {
			m["max"] = *c.Max
		}
		if len(c.Enum) > 0 {
			m["enum"] = c.Enum
		}
		props[name] = m
	}

	return map[string]any{
		"type":       "object",
		"properties": props,
	}
}

File: ts/src/sdto/SchemaBuilder.ts
import { FieldSpec } from "./FieldSpec";

// Supported sign types
export const SupportedSignTypes = ["ed25519", "rsa", "ecdsa"] as const;

// SchemaBuilder is the fluent builder for defining validation schemas
export class SchemaBuilder {
  private actions: Record<string, FieldSpec> = {};

  // SetActionStringLength sets string length constraints
  setActionStringLength(action: string, min: string, max: string): this {
    this.actions[action] = {
      type: "string",
      min,
      max,
    };
    return this;
  }

  // SetActionNumberRange sets number range constraints
  setActionNumberRange(action: string, min: string, max: string): this {
    this.actions[action] = {
      type: "number",
      min,
      max,
    };
    return this;
  }

  // SetActionEnum sets enum constraints
  setActionEnum(action: string, values: string[]): this {
    this.actions[action] = {
      type: "string",
      enum: values,
    };
    return this;
  }

  // SetActionSign sets sign field with specified sign algorithm type
  setActionSign(action: string, signType: string): this {
    this.actions[action] = {
      type: "sign",
      enum: [signType],
    };
    return this;
  }

  // SetActionSignMulti sets sign field allowing multiple sign types
  setActionSignMulti(action: string, signTypes: string[]): this {
    this.actions[action] = {
      type: "sign",
      enum: signTypes,
    };
    return this;
  }

  // BuildSchema returns the FieldSpec map for constructor use
  buildSchema(): Record<string, FieldSpec> {
    return { ...this.actions };
  }

  // Build returns JSON-friendly format for cross-service transport
  build(): Record<string, unknown> {
    const props: Record<string, unknown> = {};

    for (const [name, spec] of Object.entries(this.actions)) {
      const m: Record<string, unknown> = { type: spec.type };
      if (spec.min !== undefined) m["min"] = spec.min;
      if (spec.max !== undefined) m["max"] = spec.max;
      if (spec.enum && spec.enum.length > 0) m["enum"] = spec.enum;
      props[name] = m;
    }

    return {
      type: "object",
      properties: props,
    };
  }
}

// Factory function to match Go's NewSchemaBuilder
export function newSchemaBuilder(): SchemaBuilder {
  return new SchemaBuilder();
}
